core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -750,6 +752,10 @@ private[spark] class TaskSetManager(
       if (tasksSuccessful == numTasks) {
         isZombie = true
       }
+    } else if (fetchFailedTaskIndexSet.contains(index)) {
+      logInfo("Ignoring task-finished event for " + info.id + " in stage " + taskSet.id +
+        " because task " + index + " has already failed by FetchFailed")
+      return
```

Comment:
We can not simply `return` here. And we should always send a task `CompletionEvent` to DAG, in case of there's any listeners are waiting for it.

Reply:
Maybe, we can mark task as`FAILED` with `UnknownReason` here. And then, DAG will treat this task as no-op, and `registerMapOutput` will not be triggered. Though, it is not a elegant way.

Reply:
Yep, as @cloud-fan 's suggestion, handle this in `DAGScheduler` is a better choice.